### ISSUE TITLE
PGM no sound on big-endian

### ISSUE THREAD

**Issue by vaguerant** [libretro/FBNeo#271]
Testing the same ROM (`dmnfrnt.zip`) and BIOS (`pgm.zip`) set on a Raspberry Pi works, but on Wii U (RetroArch v1.8.4), the game locks up at the date screen.

There seems to be some kind of hardware check going on here, as the date starts in a red window on both platforms. On Pi, it turns green and the game continues, but on Wii U, the window turns yellow and the following text appears at the bottom of the screen:
`WD2LIUCK68CMD.C 151 UWRAMVAL1==0X5A5A`

Baseless speculation: some issue with the Wii U's big endian architecture.

For whatever it's worth, the PGM BIOS screen and date screen clock in at between 32 and 36 FPS on Wii U, so the game probably wouldn't run at a playable speed if it did work, making this a pretty low priority issue.

**Comment by barbudreadmon** [comment]
Yeah, our pgm driver doen't seem to be running properly on big endian platform, i saw a similar report some time ago (https://forums.libretro.com/t/cannot-play-large-pgm-games-on-wiiu/25296).

While the WiiU won't be able to run those games full speed anyway, i think PS3 (which is big endian too) should be ok. It would be nice to get confirmation that PS3 has the same issue (maybe @angomania can help with this ?)

**Comment by iosadosd-quarto** [comment]
I don't know if it's related, but with the latest DAT, Demon Front doesn't work on Windows x64 (RetroArch 1.8.5 and 1.8.6, Vulkan video driver, latest core version - updated today 05/15) and it crashes RetroArch.
Here's the log file if it can help:
[retroarch-log.txt](https://github.com/libretro/FBNeo/files/4636590/retroarch-log.txt)

Don't know why it says that it can't locate pgm archive anywhere, I build the set in full non-merged mode using the latest DAT, so all files and BIOS should be inside dmnfrnt.7z

**Comment by barbudreadmon** [comment]
> I don't know if it's related

It's not, and i can't reproduce your issue here, is that all the logs you get before crashing ?

**Comment by iosadosd-quarto** [comment]
@barbudreadmon if it's not related I'll open another issue, I initially came here searching for issues regarding Demon Front, then I saw the game and "PGM hardware" in the title, and I thought maybe my issue could be somewhat related 😅

Anyway yes, those are all the logs I get before crashing.

Maybe before opening a new issue I'll try making a split set instead of a full non-merged set, or compress the game to .zip instead of .7z, even though I never had problems with either the .7z format or the games being full non-merged.

**Comment by barbudreadmon** [comment]
@iosadosd-quarto It seems to crash between romset loading and system ram exposing (for cheevos), tbh i don't see anything that could go wrong in that part of the code, so i suspect your 7z file is just corrupted somehow (i've seen reports like this, where all files were detected but actually couldn't be copied into memory), a good test is to make sure you can actually extract your file.

I added some verbosity to the core a few minutes ago, so if it's not a corruption problem, try the newest core (when it'll be on the buildbot) and give me the new logs.

**Comment by iosadosd-quarto** [comment]
Ok so apparently FBNeo didn't like that particular 7zip. Here's what I did:

- I tried extracting the archive and all the files were extracted;
- I put those exact files in a new "dmnfrnt**.zip**" and the game ran flawlessly;
- I put the same files in a new "dmnfrnt**.7z**" and the game ran as well;
- I tried torrentzipping the .7z file (I did that to the archive that's been crashing RetroArch): the game still runs.

What baffles me is: after using TZipUI, the two 7z's now have the same SHA-1 hash, but the first one is *still* crashing RetroArch, the other one just works!

**Comment by barbudreadmon** [comment]
ok, i guess that means the 7z was indeed somehow corrupted, i know it happens sometimes, thanks for the heads-up

**Comment by iosadosd-quarto** [comment]
> ok, i guess that means the 7z was indeed somehow corrupted, i know it happens sometimes, thanks for the heads-up

Can an archive be corrupted even though I can extract all the files and use them to create a new, _working_, archive with the exact same SHA-1? I'm not doubting your conclusion (at this point, any guess is good), I'm genuinely curious

**Comment by barbudreadmon** [comment]
tbh, i don't know, i'm no archive expert, but that's the only conclusion i have since repacking solved the issue.

**Comment by dinkc64** [comment]
torrentzip adds some sort of weird blank-entries to 7zip.  We had to specifically skip them in order to not crash w/fbn standalone.  I suspect the same thing is going on with RetroArch.

This commit is what fixed the crash on our side, it basically checks the filename length, if its 0, skip the file when loading up a zip/7z file list: https://github.com/finalburnneo/FBNeo/commit/c84c9297837187b2294b0be2b91fe2d3f868d078#diff-af2207469fa08b0d78fd73e0ffbecbf0

**Comment by barbudreadmon** [comment]
@iosadosd-quarto I'm getting confused here, is the torrent7zipped file still not loading ? I thought it did after you tried again ?

**Comment by iosadosd-quarto** [comment]
> @iosadosd-quarto I'm getting confused here, is the torrent7zipped file still not loading ? I thought it did after you tried again ?

Yeah, it did work after repacking the game files, even after Torrent7zipping it

**Comment by barbudreadmon** [comment]
Ok, well, i backported https://github.com/finalburnneo/FBNeo/commit/c84c9297837187b2294b0be2b91fe2d3f868d078#diff-af2207469fa08b0d78fd73e0ffbecbf0 anyway

**Comment by dinkc64** [comment]
Repacking probably got rid of those null entries :)  either way, it's good to have this fix in the code.

**Comment by dinkc64** [comment]
tlr, does demon front work on wiiu now, or do we need some endian fixes?

**Comment by vaguerant** [comment]
@dinkc64 I just tested the latest build (64b5c46) on Wii U and Demon Front remains unchanged from the original report.

**Comment by barbudreadmon** [comment]
@vaguerant could you confirm you get the same screen on wiiu ?

![dmnfrnt](https://user-images.githubusercontent.com/10391265/82725322-7ea98b00-9cdc-11ea-93b7-a25b9fca4aa8.png)

It's important to be exhaustive when reporting that kind of issue, in this case the msg at bottom might be an important lead about the issue (this ss was taken on a ps3 emulator)

**Comment by vaguerant** [comment]
That is exactly the screen I get on Wii U, yes. My original report actually includes that debug error notice in the text:

> on Wii U, the window turns yellow and the following text appears at the bottom of the screen:
`WD2LIUCK68CMD.C 151 UWRAMVAL1==0X5A5A`

**Comment by barbudreadmon** [comment]
Oops, didn't see that part of the issue, anyway i would recommend posting a ss whenever it's possible, that's more talkative

**Comment by dinkc64** [comment]
I'll try to fix pgm soon, thanks for the confirmation.

**Comment by barbudreadmon** [comment]
demon front (and probably other pgm games) fixed on big endian hardware by https://github.com/finalburnneo/FBNeo/commit/9814ba7665c77384df428e58782f9bd4ccf880fa

**Comment by barbudreadmon** [comment]
re-opening because sound doesn't work

**Comment by barbudreadmon** [comment]
@dinkc64 i found out what's happening : calls to `ics2115write` are going bonkers (different offset & data when compared to x86), which leads to everything going bonkers : `ics2115_voice::process_sample` is not true, so `ics2115_fill_output` is not called, so `buffer` stays empty. It starts by writing 0x81 at offset 3 instead of 0x01 (on x86) every 15 calls or so, then after some time it's getting worse.

NB : i had to revert https://github.com/finalburnneo/FBNeo/commit/6afe69bd521b36eab9064f4f6be6c89995dc454f, it makes things worse : `ics2115write` isn't called at all

**Comment by barbudreadmon** [comment]
i went a bit farther, the issue seems to be about the m68k not sending everything it should to `PgmZ80WriteWord`, ending up with `RamZ80` having partial content. Something in pecular : on x86 it's writing the word `0x7676` periodically from address `0xc10000` to address `0xc1fffe`, but in the case of big endian it's doing it only once at boot

**Comment by barbudreadmon** [comment]
@dinkc64 it turns out i was right
```
		m_voice[i].vol_ctrl.value = 1;
		bprintf(PRINT_ERROR, "ics2115_reset %x %x | %x %x %x %x %x %x %x %x\n", 
			i, m_voice[i].vol_ctrl.value, 
			m_voice[i].vol_ctrl.bitflags.done, 
			m_voice[i].vol_ctrl.bitflags.stop, 
			m_voice[i].vol_ctrl.bitflags.rollover, 
			m_voice[i].vol_ctrl.bitflags.loop, 
			m_voice[i].vol_ctrl.bitflags.loop_bidir, 
			m_voice[i].vol_ctrl.bitflags.irq, 
			m_voice[i].vol_ctrl.bitflags.invert, 
			m_voice[i].vol_ctrl.bitflags.irq_pending);
```

=>

```
ics2115_reset 0 1 | 0 0 0 0 0 0 0 1
```

`>_<`

I still don't know how i'm supposed to handle this though...

**Comment by barbudreadmon** [comment]
@vaguerant It seems fixed with https://github.com/libretro/FBNeo/commit/a18eb560db402923201cb2a70c7cf6901ca89131 (please test)

**Comment by vaguerant** [comment]
Confirming all working on Wii U here, so it seems you've done it. Congrats!
